Add/stats referrer main chart

psealock opened the pull request:
Add a Main Chart to the Referrers page

### Work in Progress
![screen shot 2018-04-10 at 3 36 13 pm](https://user-images.githubusercontent.com/1922453/38534911-f4c669da-3cd4-11e8-8d88-aceebb0fb68a.png)

psealock commented:
@greenafrican @timmyc 
Could you give this a spin? How do you feel about the nested values?

**Caveats** - I'm holding off on these because I don't like the nestedValues just yet
* The legend won't update
* Tooltips not yet done

**Also**
* When no referrer is selected, I'm not sure just yet what to show. Any suggestions?
* Known issue: https://github.com/Automattic/wp-calypso/issues/24022

![screen shot 2018-04-11 at 12 25 04 pm](https://user-images.githubusercontent.com/1922453/38590219-f2ee3fb6-3d83-11e8-9671-43855c300962.png)

timmyc commented:
@psealock this is looking really cool! So at first I thought things were broken, then I saw:

> When no referrer is selected, I'm not sure just yet what to show. Any suggestions?

I think in the instance where there is data to be viewed, we just auto-select the first item in the table. So default view right now looks like:

![broken-chart](https://user-images.githubusercontent.com/22080/38694256-fff0fe00-3e3d-11e8-8091-d18b74e48471.png)

While I think we should just do this:

![should-be](https://user-images.githubusercontent.com/22080/38694292-1cd72f3a-3e3e-11e8-806b-abe909c6717e.png)

And we also need to consider what to do when there is no data for a date range at all:

![no-data](https://user-images.githubusercontent.com/22080/38694452-87afdc44-3e3e-11e8-83bb-d57286616e13.png)

Seems like we could re-use the pattern shown on the post details page when a post hasn't been viewed in the recent period:

![no-activity-post](https://user-images.githubusercontent.com/22080/38694535-c6e418ee-3e3e-11e8-9935-1049f2091546.png)

but /cc @jameskoster @kellychoffman and @LevinMedia for more thoughts

psealock commented:
### Update
Theres a lot going on here in a major refactor because the orders chart had duplicate code
* `store-stats-chart` is now a base component for the chart and handles bar selection, tab selection, rendering a legend, and legend toggling
* `formatValue` util is now using i18n's `numberFormat` function
* Normalized the way tabs and grabbed from `constants` and rendered
* Introduced selector `get-store-referrers-by-referrer.js` to filter out data for a specific referrer and change the shape of the data to match what `buildChartData` expects
* change `tabs.jsx` to accept a `format` prop so that a function can be optionally passed in. This is useful for rendering currencies instead of numbers
* Orders chart was using `endSelectedDate` and referrers was using `unitSelectedDate`. This caused errors, so I changed all the names and usages to`endSelectedDate`
* Renamed `getStoreReferrers` to `getStoreReferrersByDate` for clarity
* Automatically select the first referrer when there is none selected

### Review
This one is ready for review after the following known issues
- [ ] Handle "no data for a date range at all"

greenafrican commented:
Looking at this today (17th April) but noticed we're showing a placeholder for tomorrow (18th April).
 
![screen shot 2018-04-17 at 09 41 03](https://user-images.githubusercontent.com/1160732/38855413-ad1a6e48-4223-11e8-80be-ed9ea5183155.png)

> Automatically select the first referrer when there is none selected

Good call 👍 

I really like how nicely the search, referrer select and date selectors all work so seamlessly together. My only concern is that it isn't obvious which referrer's data is being shown in the main chart until I start playing around with the selectors. Perhaps, in another issue/PR we could emphasize which referrer the user is looking at? Perhaps something like p1523955647000027-slack-store-stats?

greenafrican commented on client/extensions/woocommerce/app/store-stats/referrers/chart/index.js:
```diff
@@ -0,0 +1,81 @@
+		return value => formatValue( value, tab.type, item.currency );
+	};
+
+	renderTabs = ( { chartData, selectedIndex, selectedTabIndex, tabClick } ) => {
```
Curious as to why `renderTabs` is here and not in `StoreStatsChart`?

psealock replied:
Good question. The orders chart has tabs that include `delta` widgets and are constructed differently than the referrer tabs. 

I could move both `renderTabs` functions to `StoreStatsChart` and conditionally apply. Would that be more clear?

greenafrican commented on client/extensions/woocommerce/app/store-stats/referrers/index.js:
```diff
@@ -124,8 +133,26 @@ class Referrers extends Component {
+				<Module
+					className="referrers__chart"
+					siteId={ siteId }
+					emptyMessage={ translate( 'No data found' ) }
```
Perhaps we can add a single "no data found"/"no activity"/"No referral activity on this date" etc. constant to client/extensions/woocommerce/app/store-stats/constants.js to ensure the string is used consistently?

greenafrican commented on client/extensions/woocommerce/app/store-stats/utils.js:
```diff
@@ -148,14 +148,15 @@ export function getStartPeriod( date, unit ) {
 	switch ( format ) {
 		case 'currency':
-			return formatCurrency( value, code );
+			return formatCurrency( value, code, { precision: decimals || 2 } );
```
Should we check for the `precision` parameters here - https://github.com/Automattic/wp-calypso/blob/a70587157b33783ba4d5e0e4a1e136680d459d51/client/lib/format-currency/index.js#L11 and https://github.com/Automattic/wp-calypso/blob/b855b7a8a7e9f6a24e4fc87e8fe251cfd10603fd/client/lib/format-currency/currencies.js#L7

psealock replied:
Good find. We are not using a specified precision with currencies and I think its better to use the defaults found, so I'm going to remove this logic

kellychoffman commented:
> I think in the instance where there is data to be viewed, we just auto-select the first item in the table. 

+1

kellychoffman commented:
When hovering over a range with 0 data, this tooltip pops up: 

![referrers](https://user-images.githubusercontent.com/5835847/38937571-babfc648-42e0-11e8-8ba3-8d820ced7e4a.png)

While, this is what pops up when hovering over a range with 0 data elsewhere in stats:

![elsewhere](https://user-images.githubusercontent.com/5835847/38937600-ce4fb484-42e0-11e8-9584-06b5f2aa94bf.png)

These should be consistent. (ie. add the value in)

psealock commented:
>Looking at this today (17th April) but noticed we're showing a placeholder for tomorrow (18th April).

@greenafrican Thats actually March 18 (see tooltip in the image you posted). That issue documented here https://github.com/Automattic/wp-calypso/issues/24022

psealock commented:
>Yes, let's keep it consistent and use the "No activity this period" pattern.

Good point, I've made that change. 

<img width="553" alt="screen shot 2018-04-18 at 2 10 12 pm" src="https://user-images.githubusercontent.com/1922453/38964403-5f58f46e-4312-11e8-868e-a01cf2f3a603.png">

>I think in the instance where there is data to be viewed, we just auto-select the first item in the table.

What about when no data is present on the selected day? The table is populated by top referrers for today (or selected day). If no referrers exist today, then we can't select anything. The chart will say "No activity this period" but that might not be entirely true. There may be referrers logged in previous days and the user has no way of selecting them or selecting a different day because the no-activity message disables day selection.

**Solution 1** - Provide some kind of UI for selection of other referrers present over entire timeframe. Crude interpretation here

<img width="533" alt="screen shot 2018-04-18 at 2 24 12 pm" src="https://user-images.githubusercontent.com/1922453/38964738-35615a14-4314-11e8-99ff-839686b7d9c1.png">

**Solution 2** - Instead of auto-selecting the first item in the table, we could show the summation of all sales and events from all referrers over the entire period in view. If today has no data, but previous days do, the user can still select previous days to see top referrers from that day. From there, they will be able to select a referrer

kellychoffman commented:
> What about when no data is present on the selected day? The table is populated by top referrers for today (or selected day). If no referrers exist today, then we can't select anything. The chart will say "No activity this period" but that might not be entirely true. There may be referrers logged in previous days and the user has no way of selecting them or selecting a different day because the no-activity message disables day selection.

The "no activity for this period" should only show if there's literally no bar graphs to show in that graph. We shouldn't add it if there's zero for the day selected.

psealock commented:
>The "no activity for this period" should only show if there's literally no bar graphs to show in that graph.

Right, for sure. The question is, since selection is based on referrers of the selected day, how do we deal with referrers who are not represented on the selected day, but might be in other days? And further, if there are zero referrers on the selected day, but there are on other days. I'm trying to figure out how to surface referrers present in the data, but not selected day.

psealock commented:
My second solution in https://github.com/Automattic/wp-calypso/pull/24021#issuecomment-382573428 seems like it could work. We would have to have really clear indications of whats being shown, however

kellychoffman commented:
Not sure I understand what Solution 2 would look like.

psealock commented:
>Not sure I understand what Solution 2 would look like.

When no referrer is selected, we show an "All Referrers" view, which is the same view except that all stats are the sum of referrers. 

For example, the chart below would be for a store with only 2 referrers logged on March 17. Their combined totals are displayed in the tabs of the chart ($300). The user can select either referrer to view specifically or select another day to see referrers logged on that day.

<img width="572" alt="screen shot 2018-04-23 at 7 33 01 am" src="https://user-images.githubusercontent.com/1922453/39143144-a2c37cba-46c8-11e8-8422-ac32f60a5dc2.png">

@kellychoffman Does this make sense?

psealock commented:
Another thing to note, @kellychoffman. If we introduce a concept of "all referrers", we'd need to include a button or way for the user to get to that state by removing selection.

timmyc commented:
> When no referrer is selected, we show an "All Referrers" view, which is the same view except that all stats are the sum of referrers.

Seems reasonable and makes sense to me 👍

kellychoffman commented:
Yes, that makes sense. I thought the issue was what to do when there's zero referrers for the current selected day (in this case)

timmyc commented:
🎉
